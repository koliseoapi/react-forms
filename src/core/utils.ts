import { RefObject, CSSProperties } from "react";

export function isNullOrUndefined(value: any): boolean {
  return typeof value === "undefined" || value === null;
}
export function isBlank(str: string | null | undefined): boolean {
  return !str || /^\s*$/.test(str);
}

export function isFalse(value: boolean | null | undefined): boolean {
  return value === false;
}

/**
 * Sets a nested property such as "foo.bar.baz", initializing intermediate nodes if needed
 */
export function setNestedProperty(root: any, path: string, value: any): any {
  let node = root;
  const pathParts = path.split(".");
  const length = pathParts.length;
  for (let i = 0; i < length; i++) {
    const [, attr, , arrayIndexStr] =
      /([^[]+)(\[(\d+)\])?/.exec(pathParts[i]) || [];
    if (!arrayIndexStr) {
      if (i == length - 1) {
        node[attr] = value;
      } else if (!node[attr]) {
        node[attr] = {};
      }
      node = node[attr];
    } else {
      if (!node[attr]) {
        node[attr] = [];
      }
      const array = node[attr];
      const arrayIndex = +arrayIndexStr;
      if (i == length - 1) {
        array[arrayIndex] = value;
      } else if (!array[arrayIndex]) {
        array[arrayIndex] = {};
      }
      node = array[arrayIndex];
    }
  }
  return root;
}

export function getNestedProperty(root: any, path: string): any {
  let node = root;
  const pathParts = path.split(/[.[\]]/).filter((part) => !!part);
  for (let part of pathParts) {
    node = node[part];
    if (typeof node === "undefined") {
      return undefined;
    }
  }
  return node;
}

/**
 * Calculate the style sto apply to an error component, to be placed just below the input element
 */
export function errorStyles(
  ref: RefObject<HTMLInputElement>
): CSSProperties | undefined {
  if (!ref.current) {
    return undefined;
  }
  const clientRect = ref.current.getBoundingClientRect();
  const parentTop = ref.current.offsetParent?.getBoundingClientRect().top || 0;
  return {
    top: clientRect.top + clientRect.height - parentTop + 8 + "px",
  };
}
