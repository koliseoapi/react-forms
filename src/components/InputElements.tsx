import React, {
  ChangeEvent,
  InputHTMLAttributes,
  useContext,
  useEffect,
  useRef,
} from "react";
import { Converter, Converters } from "../core/Converters";
import { ValidationResult } from "../core/ValidationActions";
import { errorStyles, isNullOrUndefined } from "../core/utils";
import { FormContext, FormContextContent } from "./Form";

export interface BoundComponentProps<Type>
  extends Omit<
    InputHTMLAttributes<HTMLInputElement>,
    "type" | "onChange" | "value"
  > {
  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<Type>;

  /** Optional: the validator to use. If set, all default validators (required, number, etc) will be replaced by this. Received the value after being converted */
  validate?(value: Type): Promise<ValidationResult>;

  /** onChange event handler has been extended to also receive the formContext */
  onChange?(
    event: ChangeEvent<HTMLInputElement>,
    formContext: FormContextContent
  ): void;

  /** to be used together with value, sends the converted value when the value changes */
  setValue?(value: Type | undefined): void;

  /** Optional value, for controlled components. It gets converted before calling onPropertyChange */
  value?: Type;
}

export interface InputProps<Type> extends BoundComponentProps<Type> {
  type:
    | "text"
    | "url"
    | "email"
    | "number"
    | "checkbox"
    | "date"
    | "time"
    | "radio"
    | "datetime-local";
}

export interface BoundComponentPropsWithElement<Type>
  extends BoundComponentProps<Type> {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent<Type>(
  originalProps: BoundComponentPropsWithElement<Type>
) {
  let {
    elementName,
    converter,
    validate,
    name,
    id,
    onChange: originalOnChange,
    children,
    setValue,
    value,
    ...props
  } = originalProps;
  const type: string = (props as any).type;
  if (process.env.NODE_ENV !== "production") {
    if (type === "radio" && typeof props.defaultValue === "undefined") {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
  }
  const controlled = "value" in originalProps || "setValue" in originalProps;
  if (controlled) {
    if (isNullOrUndefined(value) || !setValue) {
      throw new Error(
        `Both value and setValue are required for controlled components`
      );
    }
  }

  if (!id) {
    id = name.replace(/\./g, "-");
    if (type === "radio") {
      id = `${name}_${props.defaultValue}`;
    }
  }
  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const optionalProps: Partial<InputHTMLAttributes<HTMLInputElement>> = {};
  if (type == "checkbox") {
    optionalProps.defaultChecked = formContext.getValue(name);
  } else if (type == "radio") {
    optionalProps.defaultValue = props.defaultValue;
    optionalProps.defaultChecked =
      formContext.getValue(name) == props.defaultValue;
  } else {
    if (controlled) {
      optionalProps.value =
        "value" in originalProps
          ? converter.toValue(value)
          : converter.toValue(formContext.getValue(name));
    } else {
      optionalProps.defaultValue = converter.toValue(
        formContext.getValue(name)
      );
    }
  }

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter!.fromValue({ ...props, value: element.value });
    formContext.setValue(name, objectValue);
    setValue && setValue(objectValue as Type);
    originalOnChange && originalOnChange(e, formContext);
  }

  useEffect(() => {
    formContext.addValidator({ name, validate, converter, ...props });
    return () => formContext.removeValidator(name);
  }, []);

  const error = formContext.errors[name];

  // ARIA attributes to point to the error message
  const errorMessageId = error ? `${id}_error` : undefined;
  const adb = props["aria-describedby"];
  const ariaProps = !error
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };
  const ref = useRef(null);

  // submit on Ctrl + Enter
  function onKeyDown(e: KeyboardEvent) {
    if (e.key === "Enter" && e.ctrlKey) {
      formContext.submit();
    }
  }

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          ref,
          onKeyDown,
          ...optionalProps,
          ...props,
          ...ariaProps,
        },
        children
      )}
      {error ? (
        <div
          className="input-error"
          id={errorMessageId}
          role="alert"
          style={errorStyles(ref)}
        >
          {error}
        </div>
      ) : undefined}
    </>
  );
}

export function Input<Type>({
  autoComplete = "off",
  ...props
}: InputProps<Type>) {
  return (
    <BoundComponent
      elementName="input"
      autoComplete={autoComplete}
      {...props}
    />
  );
}

export function Select<Type>(props: BoundComponentProps<Type>) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea<Type>(props: BoundComponentProps<Type>) {
  return <BoundComponent elementName="textarea" {...props} />;
}
