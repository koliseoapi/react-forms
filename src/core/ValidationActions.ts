import { BoundComponentProps } from "../components/InputElements";
import { isBlank, isFalse, isNullOrUndefined } from "./utils";

let re_weburl: RegExp;

export type ValidationResult = string | undefined;

export type ValidationAction<Type> = (
  value: Type,
  props: BoundComponentProps<Type>
) => Promise<ValidationResult>;

async function timestampMinValidator(
  value: string,
  props: BoundComponentProps<string>
): Promise<ValidationResult> {
  // under ISO8601, dates and times can be compared as strings
  if (!isNullOrUndefined(value) && value < props.min!) {
    return "min";
  }
}

async function timestampMaxValidator(
  value: string,
  props: BoundComponentProps<string>
): Promise<ValidationResult> {
  if (!isNullOrUndefined(value) && value > props.max!) {
    return "max";
  }
}

interface IValidationActions {
  [key: string]: ValidationAction<any>;
}

export const ValidationActions: IValidationActions = {
  date_min: timestampMinValidator,
  date_max: timestampMaxValidator,
  time_min: timestampMinValidator,
  time_max: timestampMaxValidator,
  "datetime-local_min": timestampMinValidator,
  "datetime-local_max": timestampMaxValidator,

  async required(
    value: any,
    props: BoundComponentProps<any>
  ): Promise<ValidationResult> {
    if (isBlank(value) && !isFalse(props.required)) {
      return "required";
    }
  },

  async number_required(
    value: number,
    props: BoundComponentProps<number>
  ): Promise<ValidationResult> {
    if (isNullOrUndefined(value) && !isFalse(props.required)) {
      return "required";
    }
  },

  async number_min(
    value: number,
    props: BoundComponentProps<number>
  ): Promise<ValidationResult> {
    if (!isNullOrUndefined(value) && value < props.converter!.fromValue({value: '' + props.min})!) {
      return "min";
    }
  },

  async number_max(
    value: number,
    props: BoundComponentProps<number>
  ): Promise<ValidationResult> {
    if (!isNullOrUndefined(value) && value > props.converter!.fromValue({ value: '' + props.max!})!) {
      return "max";
    }
  },

  async url(
    value: string,
    props: BoundComponentProps<string>
  ): Promise<ValidationResult> {
    // Lazy init a reasonable (< 5k chars) implementation of URL regex
    // https://mathiasbynens.be/demo/url-regex
    // https://gist.github.com/dperini/729294
    if (!re_weburl) {
      re_weburl = new RegExp(
        "^" +
          // protocol identifier
          "(?:(?:https?|ftp)://)" +
          // user:pass authentication
          "(?:\\S+(?::\\S*)?@)?" +
          "(?:" +
          // IP address exclusion
          // private & local networks
          "(?!(?:10|127)(?:\\.\\d{1,3}){3})" +
          "(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})" +
          "(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})" +
          // IP address dotted notation octets
          // excludes loopback network 0.0.0.0
          // excludes reserved space >= 224.0.0.0
          // excludes network & broacast addresses
          // (first & last IP address of each class)
          "(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])" +
          "(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}" +
          "(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))" +
          "|" +
          // host name
          "(?:(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)" +
          // domain name
          "(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)*" +
          // TLD identifier
          "(?:\\.(?:[a-z\\u00a1-\\uffff]{2,}))" +
          // TLD may end with dot
          "\\.?" +
          ")" +
          // port number
          "(?::\\d{2,5})?" +
          // resource path
          "(?:[/?#]\\S*)?" +
          "$",
        "i"
      );
    }

    if (!isNullOrUndefined(value) && !re_weburl.test(value)) {
      return "url";
    }
  },

  async email(
    value: string,
    props: BoundComponentProps<string>
  ): Promise<ValidationResult> {
    // Simple email validation
    // https://emailregex.com/
    if (
      !isNullOrUndefined(value) &&
      !/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(
        value
      )
    ) {
      return "email";
    }
  },

  async pattern(
    value: string,
    props: BoundComponentProps<string>
  ): Promise<ValidationResult> {
    let { pattern } = props;
    if (!isNullOrUndefined(value) && pattern) {
      if (pattern[0] !== "^") {
        pattern = "^" + pattern;
      }
      if (pattern[pattern.length - 1] !== "$") {
        pattern += "$";
      }
      if (!new RegExp(pattern).test(value)) {
        return "pattern";
      }
    }
  },

  async maxLength(
    value: string,
    props: BoundComponentProps<string>
  ): Promise<ValidationResult> {
    if (!isNullOrUndefined(value) && value.length > props.maxLength!) {
      return "maxLength";
    }
  },
};
/**
 * Return a subset of actions that are applicable to a BoundComponent, given its properties
 * For example, for an element with [required, type=number, min=5] this method would return
 * [required, number_min]
 */
export function filterActionsForProps({
  type,
  ...props
}: any): ValidationAction<any>[] {
  const result: ValidationAction<any>[] = [];
  result.push(ValidationActions[type]);
  for (const prop of ["required", "min", "max", "pattern", "maxLength"]) {
    if (prop in props) {
      result.push(
        ValidationActions[`${type}_${prop}`] || ValidationActions[prop]
      );
    }
  }
  return result.filter((action) => !!action);
}
