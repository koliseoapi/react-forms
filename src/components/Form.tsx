import React, {
  FieldsetHTMLAttributes,
  FormEvent,
  FormHTMLAttributes,
  ReactEventHandler,
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import { Messages } from "../core/Messages";
import { getNestedProperty, setNestedProperty } from "../core/utils";
import { ValidationResult } from "../core/ValidationActions";
import { Validator, Validators } from "../core/Validators";
import { I18nContext } from "./I18nContext";
import { BoundComponentProps } from "./InputElements";

interface FormProps extends FormHTMLAttributes<HTMLFormElement> {
  /** the object being edited  */
  initialValues: any;

  /** the method to invoke when submitting, after passing all validations */
  onSubmit: (values: any) => Promise<any>;
}

/**
 * The list of all validation errors. The key is the propertyName,
 * the value is the error message
 */
export interface ValidationErrors {
  [propertyName: string]: string;
}

/**
 * Context to share with child BoundComponents
 */
export interface FormContextContent {
  /**
   * All error messages, indexed by property name
   */
  errors: ValidationErrors;

  /**
   * True if this form is submitting the value
   */
  submitting: boolean;

  /**
   * Add a Validator to the list of fields validated by this form
   */
  addValidator(props: BoundComponentProps): void;

  /**
   * Remove a validator
   */
  removeValidator(name: string): void;

  /**
   * Triggered by a change in the form field
   */
  setValue(propertyName: string, value: any): void;

  /**
   * Get value associated to a BoundComponent
   */
  getValue(propertyName: string): any;

  submit(): void;
}

function undefinedFormContext(): never {
  throw new Error("A <Form> context is required");
}
export const FormContext = createContext<FormContextContent>({
  addValidator: undefinedFormContext,
  removeValidator: undefinedFormContext,
  setValue: undefinedFormContext,
  getValue: undefinedFormContext,
  errors: {},
  submitting: false,
  submit() {},
});

/**
 * Form is the required container  of BoundComponent instances.
 */
export function Form({
  initialValues,
  onSubmit,
  children,
  ...props
}: FormProps) {
  // errors to display, sorted by component key
  const [errors, setErrors] = useState<ValidationErrors>({});

  // object with the data introduced in this form
  const [values, setValues] = useState({ ...initialValues });

  // the list of validators to use for this object
  const [validators, setValidators] = useState<Validators>({});

  // the submission status
  const [submitting, setSubmitting] = useState(false);

  // should focus on the first component with error
  const [focusFirstError, setFocusFirstError] = useState(false);

  // control if this form is still mounted
  const mounted = useRef(true);

  // validation rules to apply
  const entries = useContext(I18nContext);
  const messages = new Messages(entries);

  const onSubmitHandler: ReactEventHandler = async function (e): Promise<any> {
    e.preventDefault();
    e.stopPropagation();
    const errors = await validate();
    if (!Object.keys(errors).length) {
      setSubmitting(true);
      try {
        await onSubmit(values);
      } finally {
        // could be that the form is already dismounted
        mounted.current && setSubmitting(false);
      }
    } else {
      return errors;
    }
  };

  const formContextValue: FormContextContent = {
    errors,
    submitting,
    removeValidator: function (name) {
      delete validators[name];
      setValidators(validators);
    },

    addValidator: function (props) {
      validators[props.name] = new Validator(props);
      setValidators(validators);
    },

    setValue: function (propertyName, value) {
      setNestedProperty(values, propertyName, value);
      if (errors[propertyName]) {
        delete errors[propertyName];
        setErrors({ ...errors });
      }
      setValues(values);
    },

    getValue(propertyName) {
      return getNestedProperty(values, propertyName);
    },

    submit() {
      onSubmitHandler({
        preventDefault() {},
        stopPropagation() {},
      } as FormEvent);
    },
  };

  async function validate(): Promise<ValidationErrors> {
    const newErrors: ValidationErrors = {};
    await Promise.all(
      Object.entries(validators).map(async ([propertyName, validator]) => {
        const error: ValidationResult = await validator.validate(
          getNestedProperty(values, propertyName)
        );
        if (!!error) {
          newErrors[propertyName] = messages.get(error, validator.props);
        }
      })
    );
    setErrors(newErrors);
    if (Object.keys(newErrors).length > 0) {
      setFocusFirstError(true);
    }
    return newErrors;
  }

  useEffect(() => {
    if (focusFirstError) {
      const firstInvalidElement = document.body.querySelector<HTMLInputElement>(
        "[aria-invalid=true]"
      );
      firstInvalidElement && firstInvalidElement.focus();
      setFocusFirstError(false);
    }
  }, [focusFirstError]);

  useEffect(() => {
    // React 18 in dev mode with strict mode seems to run useEffect twice
    // https://github.com/facebook/react/issues/24670#issuecomment-1280222217
    mounted.current = true;
    return () => {
      mounted.current = false;
    };
  }, []);

  return (
    <form {...props} onSubmit={onSubmitHandler} noValidate>
      <FormContext.Provider value={formContextValue}>
        {children}
      </FormContext.Provider>
    </form>
  );
}

/**
 * A fieldset that will disable all input controls while submitting
 */
export function FieldSet({
  children,
  ...props
}: FieldsetHTMLAttributes<HTMLFieldSetElement>) {
  const formContext = useContext(FormContext);
  return (
    <fieldset disabled={formContext.submitting} {...props}>
      {children}
    </fieldset>
  );
}
