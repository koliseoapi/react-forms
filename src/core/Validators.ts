import { BoundComponentProps } from "../components/InputElements";
import {
  ValidationAction,
  ValidationResult,
  filterActionsForProps,
} from "./ValidationActions";

export class Validator<Type> {
  propertyName: string;

  /** props assigned to the BoundComponent */
  props: BoundComponentProps<Type>;

  /** Validation actions to execute on this property */
  actions: ValidationAction<Type>[];

  /**
   * If props.validate is not undefined, it will replace the automatic validations.
   * Otherwise a list of validations based on type, min, max, pattern will be set.
   */
  constructor(props: BoundComponentProps<Type>) {
    this.propertyName = props.name;
    this.actions = props.validate
      ? [props.validate]
      : filterActionsForProps(props);
    this.props = props;
  }

  /**
   * Launches all validation actions against the current value
   */
  async validate(currentValue: any): Promise<ValidationResult> {
    for (const action of this.actions) {
      const error = await action(currentValue, this.props);
      if (typeof error !== "undefined") {
        return error;
      }
    }
    return undefined;
  }
}

export interface Validators {
  [propertyName: string]: Validator<any>;
}
