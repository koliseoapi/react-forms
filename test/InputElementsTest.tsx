import React, { ReactElement } from "react";
import renderer, { ReactTestRenderer, act } from "react-test-renderer";
import { Input, Select } from "../src/components/InputElements";
import { Form, ValidationErrors, FieldSet } from "../src/components/Form";
import { Converters } from "../src/core/Converters";
import { ValidationResult } from "../src/core/ValidationActions";

describe("Input", function () {
  let onSubmit: () => Promise<string>;

  let form: ReactTestRenderer;

  function mount(input: ReactElement, initialValues: any) {
    onSubmit = jest.fn(async () => "foo");
    // act() to wait for the useEffect() calls on initial render
    act(() => {
      form = renderer.create(
        <Form onSubmit={onSubmit} initialValues={initialValues}>
          {input}
        </Form>
      );
    });
  }

  function triggerChange({
    value,
    checked,
  }: {
    value: string;
    checked?: boolean;
  }) {
    act(() => {
      const inputElement = form.root.findAll((el) => el.type == "input")[0];
      expect(inputElement.type).toBe("input");
      inputElement.props.onChange({
        target: {
          value,
          checked,
        },
      });
    });
  }

  async function triggerSubmit(expectedErrors?: ValidationErrors) {
    return act(async () => {
      const formElement = form.root.find((el) => el.type == "form");
      const errors = await formElement.props.onSubmit({
        preventDefault() {},
        stopPropagation() {},
      });
      if (!errors) {
        expect(expectedErrors).toBeUndefined();
      } else {
        expect(expectedErrors).toBeDefined();
        expect(errors).toMatchObject(expectedErrors!);
      }
    });
  }

  it("should render correctly with type=text", () => {
    mount(<Input type="text" className="foo" required name="name" />, {
      name: "foobar",
    });
    // input has class 'foo'
    // input is required
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should use checked instead of value with type=checkbox", () => {
    mount(<Input type="checkbox" name="subscribed" />, { subscribed: true });
    // input[checked] is set
    // input[value] is not set
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render type=radio", () => {
    mount(
      <>
        <Input type="radio" name="abtesting" defaultValue="a" />
        <Input type="radio" name="abtesting" defaultValue="b" />
      </>,
      { abtesting: "a" }
    );
    // input[checked] is set
    // different id calculated for each input
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render type=radio", function () {
    mount(
      <>
        <Input type="radio" name="gender" defaultValue="male" />,
        <Input type="radio" name="gender" defaultValue="female" />
      </>,
      { gender: "male" }
    );
    // male selected, female unselected
    expect(form.toJSON()).toMatchSnapshot(
      "radio: male selected, female unselected"
    );

    // trigger change in input2
    const input2 = form.root.findAll((el) => el.type == "input")[1];
    input2.props.onChange({
      target: { value: "female", checked: true },
    });
    expect(form.toJSON()).toMatchSnapshot(
      "radio: female selected, male unselected"
    );
  });

  it("should not propagate specific properties, like converter, to the HTML5 element", () => {
    mount(
      <Input
        type="text"
        className="foo"
        name="name"
        converter={Converters.text}
      />,
      {}
    );
    // state attribute should not be propagated to nested <input>
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render a validation error", async function () {
    mount(<Input type="text" required name="name" />, {});
    await triggerSubmit({ name: "Please fill out this field" });
    // error message was rendered
    // error message is accessible: role=alert, aria-invalid, aria-describedby
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should add validation error to preexisting aria-describedby attribute", async function () {
    mount(
      <Input type="text" required name="name" aria-describedby="foo" />,
      {}
    );
    await triggerSubmit({ name: "Please fill out this field" });
    // error message was rendered
    // error message is accessible: role=alert, aria-invalid, aria-describedby
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should update number (also: onChange listener)", async () => {
    const values = { age: 20, color: "red" };
    const callback = jest.fn();
    mount(
      <Input type="number" name="age" step="1" min="1" onChange={callback} />,
      values
    );
    triggerChange({ value: "32.2" });
    await triggerSubmit();
    expect(onSubmit).toHaveBeenCalledWith({
      age: 32,
      color: "red",
    });
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("should update nested properties", async () => {
    const values: {
      // empty initially
      from?: { address: string };

      // not empty initially
      to?: { address: string };
    } = { to: { address: "barbaz" } };
    mount(
      <>
        <Input type="text" name="from.address" required />,
        <Input type="text" name="to.address" />,
      </>,
      values
    );
    triggerChange({ value: "foobar" });
    await triggerSubmit();
    expect(onSubmit).toHaveBeenCalledWith({
      from: { address: "foobar" },
      to: { address: "barbaz" },
    });
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("required and optional checkboxes (also: set id attribute)", async function () {
    mount(
      <Input
        type="checkbox"
        name="subscribed"
        id="foobarbaz"
        required={false}
      />,
      {
        subscribed: false,
      }
    );
    await triggerSubmit();

    mount(
      <Input
        type="checkbox"
        name="subscribed"
        id="foobarbaz"
        required={true}
      />,
      {
        subscribed: false,
      }
    );
    await triggerSubmit({ subscribed: "Please fill out this field" });
    expect(form.toJSON()).toMatchSnapshot();
  });

  function validate(value: string): Promise<ValidationResult> {
    return Promise.resolve(value === "abc" ? undefined : "pattern");
  }

  it("custom async validation fails (also: onSubmit not called)", async function () {
    mount(<Input type="text" name="foo" validate={validate} />, { foo: "xyz" });
    await triggerSubmit({ foo: "Please match the requested format" });
    expect(form.toJSON()).toMatchSnapshot();
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it("custom async validation passes (also: onSubmit called)", async function () {
    mount(<Input type="text" name="foo" validate={validate} />, { foo: "abc" });
    await triggerSubmit();
    expect(form.toJSON()).toMatchSnapshot();
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });

  it("has access to formContext from onChange", async function () {
    mount(
      <Input
        type="checkbox"
        name="foo"
        onChange={(e, formContext) => {
          const value = e.target.checked;
          expect(value).toBe(formContext.getValue("foo"));
        }}
      />,
      { foo: true }
    );
    triggerChange({ value: "xxx", checked: true });
  });
});
