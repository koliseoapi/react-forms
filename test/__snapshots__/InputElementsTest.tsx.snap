// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input custom async validation fails (also: onSubmit not called) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="foo_error"
    aria-invalid={true}
    autoComplete="off"
    defaultValue="xyz"
    id="foo"
    name="foo"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
  />
  <div
    className="input-error"
    id="foo_error"
    role="alert"
  >
    Please match the requested format
  </div>
</form>
`;

exports[`Input custom async validation passes (also: onSubmit called) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultValue="abc"
    id="foo"
    name="foo"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
  />
</form>
`;

exports[`Input required and optional checkboxes (also: set id attribute) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="foobarbaz_error"
    aria-invalid={true}
    autoComplete="off"
    defaultChecked={false}
    id="foobarbaz"
    name="subscribed"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="checkbox"
  />
  <div
    className="input-error"
    id="foobarbaz_error"
    role="alert"
  >
    Please fill out this field
  </div>
</form>
`;

exports[`Input should add validation error to preexisting aria-describedby attribute 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="name_error foo"
    aria-invalid={true}
    autoComplete="off"
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
  />
  <div
    className="input-error"
    id="name_error"
    role="alert"
  >
    Please fill out this field
  </div>
</form>
`;

exports[`Input should not propagate specific properties, like converter, to the HTML5 element 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="foo"
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
  />
</form>
`;

exports[`Input should render a validation error 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="name_error"
    aria-invalid={true}
    autoComplete="off"
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
  />
  <div
    className="input-error"
    id="name_error"
    role="alert"
  >
    Please fill out this field
  </div>
</form>
`;

exports[`Input should render correctly with type=text 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="foo"
    defaultValue="foobar"
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
  />
</form>
`;

exports[`Input should render type=radio 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultChecked={true}
    defaultValue="a"
    id="abtesting_a"
    name="abtesting"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
  <input
    autoComplete="off"
    defaultChecked={false}
    defaultValue="b"
    id="abtesting_b"
    name="abtesting"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should render type=radio: radio: female selected, male unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultChecked={true}
    defaultValue="male"
    id="gender_male"
    name="gender"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
  ,
  <input
    autoComplete="off"
    defaultChecked={false}
    defaultValue="female"
    id="gender_female"
    name="gender"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should render type=radio: radio: male selected, female unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultChecked={true}
    defaultValue="male"
    id="gender_male"
    name="gender"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
  ,
  <input
    autoComplete="off"
    defaultChecked={false}
    defaultValue="female"
    id="gender_female"
    name="gender"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should update nested properties 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultValue="foobar"
    id="from-address"
    name="from.address"
    onChange={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
  />
  ,
  <input
    autoComplete="off"
    defaultValue="barbaz"
    id="to-address"
    name="to.address"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
  />
  ,
</form>
`;

exports[`Input should use checked instead of value with type=checkbox 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    defaultChecked={true}
    id="subscribed"
    name="subscribed"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="checkbox"
  />
</form>
`;

exports[`Input should work when controlled 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    id="age"
    name="age"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="number"
    value="30"
  />
</form>
`;

exports[`Input should work when controlled 2`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    id="age"
    name="age"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="number"
    value="30"
  />
</form>
`;

exports[`Input should work when controlled 3`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    id="name"
    name="name"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</form>
`;
